src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,233 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+
+const getScrollableParent = (element: HTMLElement | null): HTMLElement => {
+  if (!element || element === document.body) {
+    return document.body;
+  }
+
+  return isElementScrollable(element)
+    ? element
+    : getScrollableParent(element.parentElement);
+};
+
+const isElementScrollable = (element: HTMLElement) => {
+  return /(auto|scroll|overlay)/.test(
+    getElementStyle(element, 'overflow') +
+      getElementStyle(element, 'overflow-y'),
+  );
+};
+
+const getElementStyle = (element: HTMLElement, prop: string) => {
+  return getComputedStyle(element, null).getPropertyValue(prop);
+};
+
+const getElementHeight = (element: HTMLElement | undefined) => {
+  return element?.getBoundingClientRect().height ?? 0;
+};
+
+const getNumberOfNodesInHeight = (childHeight: number, totalHeight: number) => {
+  if (!childHeight) {
+    return 0;
+  }
+
+  return Math.floor(totalHeight / childHeight);
+};
+
+const getTranslateValue = (childHeight: number, startIndex: number) => {
+  return childHeight * startIndex;
+};
+
+const getVisibleNodeCount = (
+  childHeight: number,
+  startIndex: number,
+  childrenLength: number,
+  scrollContainer: HTMLElement,
+) => {
+  return Math.min(
+    childrenLength - startIndex,
+    getNumberOfNodesInHeight(childHeight, getElementHeight(scrollContainer)),
+  );
+};
+
+export type VirtualScrollProps = {
+  /**
+   * Length of the items to virtualize.
+   */
+  itemsLength: number;
+  /**
+   * Single item render function, which gives index of the item (0 based) in the data array
+   * and expects to get the JSX of that element to render.
+   * Recommended to memoize the reference of the function.
+   */
+  itemRenderer: (index: number) => JSX.Element;
+  /**
+   * Number of items to be rendered at the end.
+   * Not recommended to go lower than the visible items in viewport.
+   * @default 20
+   */
+  buffer?: number;
+};
```

Comment:
Maybe add `CommonProps` too, if someone wants to add custom class or styling e.g. some container class as we are changing DOM (as you mentioned in the comments).

Reply:
added